README.md
Latest xlslib News

Join the new "xlslib-news" Mail List to get notified as soon as new versions are posted.

*** REQUEST FOR HELP ***
We have one outstanding feature request - to support Visual Studio 2010. Would some user of such give us a hand their and upload the necessary files?
************************

*** NOTES ***
The svn repository has a modified sheet class - from now on "rowheight" will be in TWIPS.
*************

This library allows a C++ or C programmer to dynamically construct Excel spreadsheets in code. Download the zip file, unzip it, cd into the new xlslib, and build the library. A few test files in the test folder demonstrate usage.

./configure
make
make check
make install

RELEASE NOTES: 

Release 2.3.4 January 15, 2012
- the "area" feature for formulas was not exposed - so exposed it. Thus can now do SUMS(area):

  expression_node_factory_t& maker = wb.GetFormulaFactory();
  cell_t *c1 = sh->number(0, 0, 1.1);
  sh->number(0, 1, 2.2);
  cell_t *c2 = sh->number(0, 2, 3.3);
  expression_node_t *pExpFormula = maker.f(FUNC_SUM, maker.area(*c1, *c2, CELL_RELATIVE_A1));
  sh->formula(0, 3, pExpFormula, true);

Release 2.3.0 January 2, 2012

- new feature: attach a HyperLink to a cell with optional '#' mark
  Can be any type of cell - number, label, etc. Also in bridge.
- new feature: can now set default values for column width and row height
- support for building the project with CodeBlock, RadStudio, and BDS
- blocks now outputted per Excel specification
- hopefully fixed the md5 checksums in "make check"
- code was reformatted with uncrustify to make it easier for me to support
- header cleanup so only required headers in the interface files
- headers in the include directory are now placed in subdirectories.
  IMPORTANT NOTE: You should uninstall the old library first using old source
  (Version 2.2.0, archive still visible), only then install the latest 2.3.0.
  Or, you can do it manually. Look at the /usr/local/include/xlslib folder
  should be directories and only one file, xlslib.h.
- one known bug - creating an empty spreadsheet with no cells is broken.
  File is corrupt according to Excel. This was broken, fixed, now broken again.

Finally: each of the tests run in make check show 100% clean when run with valgrind! You can thank Ger Hobbelt for that!

Release 2.2.0 November 11, 2011
- several bug fixes

October 22, 2011
Release 2.1.0
- support for UTF-8 strings (must have iconv on your system)
- fixed a problem where files exactly 130 512blocks long were unreadable
- hopefully fixed the md5 checksums in "make check"
- converted license from LGPL to BSD

October 16, 2011
Release 2.0.0
- now using SST in support of text LABELs with strings > 255 (8224 bytes supported)
- several bug fixes.
- significant performance improvements
- some support for Formulas (but no help right now) - see the mainC.cpp in test.

November 26, 2009
Release 1.6.0
- release includes a configure script so bootstrapping is not always required
- several bug fixes.
- significant performance improvements
- this is the last release on the original software base - version 2.0 will be quite different internally and much faster

June 21, 2009
Release 1.5.0
- huge number of bug fixes.
- performance improvements - 30x faster - will create a 64000x256 sheet in about 3 minutes.
- now enables creation of document properties
- DHxls Objective-C Framework (SVN repository, trunk/DHxls)

October 25, 2008
Release 1.2.0
- Unicode support (upgrade format to BIFF8) along with previous ASCII
- unlimited size - tested with 256 columns by 64000 rows
- additional formatting options with the new BIFF8 (look in header files)

February 9, 2008
Release 1.0.0
- buildable as 32 or 64 bit
- added a "C" interface to support C-only applications
- requires only the inclusion of a single header file:
- general code cleanup and fixes

April 25th, 2007
Release 0.2.5
- several code and makefile fixes so the project would build
- fixes to the test application so it would build
- license was changed to the Lesser GLP
September 1st, 2004
Release 0.2.1. First stable usable release.



xlslib/NEWS
xlslib NEWS
------------------
CVS ID: $Id: NEWS,v 1.3 2008/10/25 18:48:48 dhoerl Exp $

# someday
#The most up-to-date version of this file is:
#http://xlslib.cvs.sourceforge.net/xlslib/NEWS?view=markup

Date 2012-07-12
Summary: xlslib release 2.3.4 (BETA)
-----------------------------------------------------------
- MSVC2010 project files fixed
- MingW support in ./configure script and makefiles
- etc.etc.etc.

Date 2007-4-15
Summary: xlslib release 0.2.5 (ALPHA)
-----------------------------------------------------------
This is an alpha release of xlslib.
It includes:

- library source changes so it will compile
- configure script rebuilt
- test application now compiles and runs
- Various other improvements
- Its quite likely that the sister PHP library is broken,
  if anyone wants to help to get it operational, please 
  contact the administrators.
- A "C" interface is in the works - stay tuned.

Date 2008-02-09
Summary: xlslib release 1.0.0
-----------------------------------------------------------
- 32/64 bit buildable
- make clean works better
- A "C" interface - see xlslib.h for interface

Date 2008-10-1
Summary: xlslib release 1.2.0
-----------------------------------------------------------
- Updated to create BIFF8 formatted files (from BIFF5) for:
- full 256 columns x 64000 rows!
- support for ASCII and UNICODE strings in labels and sheet names
- compliance with mandatory items specified in OpenOffice.org's Excel Spec

Date 2009-6-20
Summary: xlslib release 1.5.0
-----------------------------------------------------------
- Significant bug fixes and performance improvements
- Possible now to create "Properties" objects



xlslib/src/xlslib/note.cpp
/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 *
 * This file is part of xlslib -- A multiplatform, C/C++ library
 * for dynamic generation of Excel(TM) files.
 *
 * Copyright 2004 Yeico S. A. de C. V. All Rights Reserved.
 * Copyright 2008-2011 David Hoerl All Rights Reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification, are
 * permitted provided that the following conditions are met:
 *
 *    1. Redistributions of source code must retain the above copyright notice, this list of
 *       conditions and the following disclaimer.
 *
 *    2. Redistributions in binary form must reproduce the above copyright notice, this list
 *       of conditions and the following disclaimer in the documentation and/or other materials
 *       provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY David Hoerl ''AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
 * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL David Hoerl OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */

#include "xlslib/record.h"
#include "xlslib/note.h"
#include "xlslib/globalrec.h"
#include "xlslib/datast.h"
#include "xlslib/rectypes.h"


using namespace xlslib_core;
using namespace xlslib_strings;

/*
 *********************************
 *  note_t class implementation
 *********************************
 */
note_t::note_t(CGlobalRecords& gRecords, unsigned32_t rowval, unsigned32_t colval, const std::string& msg, const std::string& auth, xf_t* pxfval) :
	cell_t(gRecords, rowval, colval, pxfval)
{
	gRecords.char2str16(msg, this->text);
	gRecords.char2str16(auth, this->author);
}

note_t::note_t(CGlobalRecords& gRecords, unsigned32_t rowval, unsigned32_t colval, const ustring& msg, const ustring& auth, xf_t* pxfval) :
	cell_t(gRecords, rowval, colval, pxfval)
{
	gRecords.wide2str16(msg, this->text);
	gRecords.wide2str16(auth, this->author);
}

#ifndef __FRAMEWORK__
note_t::note_t(CGlobalRecords& gRecords, unsigned32_t rowval, unsigned32_t colval, const u16string& msg, const u16string& auth, xf_t* pxfval) :
	cell_t(gRecords, rowval, colval, pxfval),
	text(msg),
	author(auth)
{
}

#endif

note_t::~note_t()
{
}

size_t note_t::GetSize(void) const
{
	return 12;
}

CUnit* note_t::GetData(CDataStorage &datastore) const
{
	return datastore.MakeCNote(*this);  // NOTE: this pointer HAS to be deleted elsewhere.
}

/*
 *********************************
 *  CNote class implementation
 *********************************
 *
 *  BIFF ROW (208h)  16  00 00 00 00 03 00 2C 01  00 00 00 00 00 01 0F 00
 *  BIFF 6 (06h)  43  00 00 00 00 0F 00 8C 16  22 AA FD 90 F8 3F 00 00
 *  C0 00 00 FD 15 00 1E 01  00 1F 00 00 00 00 00 00
 *  D0 3F 04 41 13 00 1E 04  00 06 03
 *  BIFF 6 (06h)  39  00 00 01 00 0F 00 8C 16  22 AA FD 90 98 40 00 00
 *  00 00 00 FE 11 00 44 00  00 00 C0 44 00 00 02 C0
 *  05 44 00 00 02 C0 05
 *  BIFF RK (27Eh)  10  00 00 02 00 0F 00 00 00  40 40
 *  BIFF DBCELL (D7h)  6  7C 00 00 00 00 00
 *  BIFF MSODRAWING (ECh)  224  0F 00 02 F0 7E 01 00 00  10 00 08 F0 08 00 00 00
 *  03 00 00 00 02 04 00 00  0F 00 03 F0 66 01 00 00
 *  0F 00 04 F0 28 00 00 00  01 00 09 F0 10 00 00 00
 *  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
 *  02 00 0A F0 08 00 00 00  00 04 00 00 05 00 00 00
 *  0F 00 04 F0 90 00 00 00  A2 0C 0A F0 08 00 00 00
 *  01 04 00 00 00 0A 00 00  D3 00 0B F0 4E 00 00 00
 *  80 00 A8 92 41 0C 8B 00  02 00 00 00 BF 00 08 00
 *  08 00 58 01 00 00 00 00  81 01 FF FF E1 00 83 01
 *  FF FF E1 00 85 01 F4 00  00 10 BF 01 10 00 10 00
 *  C3 01 F4 00 00 10 01 02  00 00 00 00 03 02 F4 00
 *  00 10 3F 02 03 00 03 00  BF 03 02 00 02 00 00 00
 *  10 F0 12 00 00 00 03 00  00 00 60 02 01 00 40 00
 *  02 00 60 02 04 00 F3 00  00 00 11 F0 00 00 00 00
 *  BIFF OBJ (5Dh)  52  15 00 12 00 19 00 01 00  11 40 A8 92 41 0C 80 8D
 *  42 0C 00 00 00 00 0D 00  16 00 D8 AC 8C FF 20 14
 *  4F 47 BA 90 F3 6F D0 9F  7B C0 00 00 10 00 00 00
 *  00 00 00 00
 *  BIFF MSODRAWING (ECh)  8  00 00 0D F0 00 00 00 00
 *  BIFF TXO (1B6h)  18  12 02 00 00 00 00 00 00  00 00 1A 00 10 00 00 00
 *  00 00
 *  BIFF CONTINUE (3Ch)  27  00 41 42 43 44 45 46 47  48 49 4A 4B 4C 4D 4E 4F
 *  50 51 52 53 54 55 56 57  58 59 5A
 *  BIFF CONTINUE (3Ch)  16  00 00 05 00 BF 00 0E 00  1A 00 00 00 00 00 00 00
 *  BIFF MSODRAWING (ECh)  150  0F 00 04 F0 96 00 00 00  A2 0C 0A F0 08 00 00 00
 *  02 04 00 00 00 0A 00 00  E3 00 0B F0 54 00 00 00
 *  80 00 28 C8 3F 04 85 00  01 00 00 00 8B 00 02 00
 *  00 00 BF 00 08 00 0A 00  58 01 00 00 00 00 81 01
 *  FF FF E1 00 83 01 FF FF  E1 00 85 01 F4 00 00 10
 *  BF 01 10 00 10 00 C3 01  F4 00 00 10 01 02 00 00
 *  00 00 03 02 F4 00 00 10  3F 02 03 00 03 00 BF 03
 *  00 00 02 00 00 00 10 F0  12 00 00 00 03 00 01 00
 *  90 01 01 00 4D 00 02 00  20 03 02 00 33 00 00 00
 *  11 F0 00 00 00 00
 *  BIFF OBJ (5Dh)  52  15 00 12 00 19 00 02 00  11 40 28 C8 3F 04 20 8A
 *  42 0C 00 00 00 00 0D 00  16 00 19 B5 48 49 52 4A
 *  AF 49 9E AB 90 6C 27 12  73 34 00 00 8B 00 02 00
 *  00 00 00 00
 *  BIFF MSODRAWING (ECh)  8  00 00 0D F0 00 00 00 00
 *  BIFF TXO (1B6h)  18  12 02 00 00 00 00 00 00  00 00 0A 00 10 00 00 00
 *  00 00
 *  BIFF CONTINUE (3Ch)  11  00 30 31 32 33 34 35 36  37 38 39
 *  BIFF CONTINUE (3Ch)  16  00 00 05 00 46 00 0E 00  0A 00 72 00 00 00 00 00
 *  BIFF NOTE (1Ch)  16  00 00 00 00 00 00 01 00  04 00 00 75 73 65 72 00
 *  BIFF NOTE (1Ch)  16  00 00 01 00 00 00 02 00  04 00 00 75 73 65 72 00
 *  BIFF WINDOW2 (23Eh)  18  B6 00 00 00 00 00 40 00  00 00 00 00 00 00 72 00
 *  00 00
 *
 */

CNote::CNote(CDataStorage &datastore, const note_t& notedef) :
	CRecord(datastore)
{
	unsigned16_t idx = 1; // OBJ reference

	SetRecordType(RECTYPE_NOTE);
	AddValue16((unsigned16_t)notedef.GetRow());
	AddValue16((unsigned16_t)notedef.GetCol());
	AddValue16(0); // grBit
	AddValue16(idx);
	AddUnicodeString(notedef.GetAuthor(), LEN2_NOFLAGS_PADDING_UNICODE);

	SetRecordLength(GetDataSize()-RECORD_HEADER_SIZE);
}

CNote::~CNote()
{
}

// make an OBJ record:
void CNote::mk_obj_Record(const note_t* notedef)
{
	unsigned16_t idx = 1;

	SetRecordType(RECTYPE_OBJ);
	AddValue16((unsigned16_t)notedef->GetRow());
	AddValue16((unsigned16_t)notedef->GetCol());
	AddValue16(0); // grBit
	AddValue16(idx);

	SetRecordLength(GetDataSize()-RECORD_HEADER_SIZE);
}

// start
void CNote::mk_obj_CMO_SubRecord(const note_t* notedef)
{
	(void)notedef; // stop warning
	AddValue16(0x15);  // ftCmo
	AddValue16(14+12-4);

	AddValue16(0x19); // ot = Comment
	AddValue16(1);    // id = OBJ id = 1
	AddValue16(0);    // flags
	AddFixedDataArray(0, 14+12-10);    // reserved

	SetRecordLength(GetDataSize()-RECORD_HEADER_SIZE);
}

// end
void CNote::mk_obj_END_SubRecord(const note_t* notedef)
{
	(void)notedef; // stop warning
	AddValue16(0x00); // ftEnd
	AddValue16(0);

	SetRecordLength(GetDataSize()-RECORD_HEADER_SIZE);
}

// note structure	???
void CNote::mk_obj_NTS_SubRecord(const note_t* notedef)
{
	(void)notedef; // stop warning
	AddValue16(0x0D); // ftNts
	AddValue16(0);

	SetRecordLength(GetDataSize()-RECORD_HEADER_SIZE);
}
